### ISSUE TITLE
minetest.wrap_text is broken

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#6234]
`minetest.wrap_test` looks very broken to me, it returns very weird result in MT 0.4.16.

I used `luacmd` to test.

This function returns a *table* of strings. As if the source string has been splitted. There are no newlines. And the strings can even be *longer* than the length limit. This output is not useful at all.
Proof:
![My mama tell me](https://i.imgur.com/J40l346.png)
https://i.imgur.com/J40l346.png

To test for yourself, just try to call `minetest.wrap_text` with help of the `luacmd` and experiment a bit. You will see what happens.

**Comment by SmallJoker** [comment]
See #6056 for possible fix.

**Comment by Tmanyo** [comment]
I don't get it...  SmallJoker you say the function works because Wuzzy2 proved it did.  1) He proved it didn't work. 2) You close that issue based on false information.  3) You comment the pull request I started stating that it is a possible fix even though you already claimed that the function worked?

**Comment by dsohler** [comment]
Wait a second. So instead of getting a wrapped string when using text_wrap on an unwrapped string I get a table? Wow, Skipped API design 101?

**Comment by SmallJoker** [comment]
@Tmanyo 
> Whenever I try to use minetest.wrap_text it does not seem to work. It always says that I am attempting to call field 'wrap_text' (a nil value).

The nil value error you described is not related to the function, as it probably was overwritten or emptied from any mod. However, the issue description could've been edited to explain the misbehaviour of this function.

@dsohler Please see https://github.com/minetest/minetest/pull/6056#issuecomment-319305505 for the explanation why it must return a table and how you can convert it to a string easily. Would be helpful to add a notice to the docs.

**Comment by dsohler** [comment]
@SmallJoker So it needs to be a table because the UI is messed up and wants a table where a string would be correct :smile:

Okay then fixing the documentation about it would be great because [lua.api.txt](https://github.com/minetest/minetest/blob/816bca32ac69f58b7de881d68689c6a1e3897a0e/doc/lua_api.txt) and [client_lua_api.txt](https://github.com/minetest/minetest/blob/816bca32ac69f58b7de881d68689c6a1e3897a0e/doc/client_lua_api.md) both state in their most recent version as of writing this comment “returns a string” and “Adds new lines to the string to keep it within the specified character limit”.

**Comment by Tmanyo** [comment]
@SmallJoker Well I guess if that is justification enough for you, haha.  That still doesn't change the fact that with this function, it should return a string.  Oh and it is a broken function.  Seems to me the UI should be fixed to work with a string...

@dsohler Stop making sense... I have tried already and it doesn't get this topic anywhere.  LOL

**Comment by rubenwardy** [comment]
really the function should be called `wrap_text_to_table` or similar

**Comment by Tmanyo** [comment]
@rubenwardy Well, then could we have a second more useful function that just wraps the text and doesn't even consider a table?

**Comment by raymoo** [comment]
I don't think it's a huge deal that it's in a table, since you can just use `table.concat`. However I feel like just turning it into text is the more common case, so there should be a function for that too. The table version should be kept though since it's cheaper to make the table than to go to a string, and it would be more expensive than that to get a string and then go back and split it into a table.

**Comment by dsohler** [comment]
> I don't think it's a huge deal that it's in a table, since you can just use `table.concat`.

We're already over this. The question now is: Why is the documentation wrong and why is the UI messed up? :smile:

**Comment by HybridDog** [comment]
https://github.com/minetest/minetest/commit/0120fe16a761f9e06c8c2877439db6a46d808143

**Comment by dsohler** [comment]
What does this has to do with the wrong documentation or wrong return type of the function in question?

**Comment by HybridDog** [comment]
The documentation seems to be added there (see bottom of the page) and the function was called minetest.splittext before.

**Comment by dsohler** [comment]
But you **do** realize that the description still says it returns a string but actually it returns a table?

**Comment by HybridDog** [comment]
Yes, l guess the documentation is wrong because of this commit.

**Comment by dsohler** [comment]
> l guess the documentation is wrong because of this commit.

It was wrong before. Now it’s consistently wrong in all places.

**Comment by Tmanyo** [comment]
Here is what the simple solution is.  If you input _**text**_ in a _**text**_ wrapping function, you should get.... wait for it.... _**text**_.  Getting a table makes no sense to me and possibly others.  I think that the documentation is great because it makes total sense.  I don't know who created the faulty function and I don't know who wrote the API description.  However, it seems that if it were two different people (I am guessing), the person who wrote the API description worded it in the way that made most sense to him/her.  The whole issue with the UI and the textlist can be solved while using a string.  I like to make formspecs with player input that can be posted on a textlist in paragraph form, never would I use a table.  Also, there needs to be a formspec element that allows to present text in a paragraph form without having to worry about commas creating new lines...

**Comment by rubenwardy** [comment]
> Also, there needs to be a formspec element that allows to present text in a paragraph form without having to worry about commas creating new lines...

You should be using minetest.formspec_escape

**Comment by Tmanyo** [comment]
> You should be using minetest.formspec_escape

I have, but I have found that it doesn't always work as I intend it to.

**Comment by rubenwardy** [comment]
> I have, but I have found that it doesn't always work as I intend it to.

Then there's a bug with that function. There's no need to add a new formspec type.

**Comment by Tmanyo** [comment]
> There's no need to add a new formspec type.

You are very correct.  There is no need to create a new formspec element.  However, it would make life so much easier. lol  IMO the formspecs need to be reworked, but I know others disagree.

**Comment by dsohler** [comment]
> I think that the documentation is great because it makes total sense.

The documentation describes what it should be. It describes what's expected when putting text into a text wrapping function. But that is not the purpose of the documentation. The documentation should document what the function actually does.

So leaving aside that the maximum length of lines (or table entries in the current state) seems to be only a rough approximate value instead of a maximum length limiter: either the function is broken or the documentation is wrong.

One or another has to be fixed.

**Comment by Tmanyo** [comment]
> either the function is broken or the documentation is wrong.

It is pretty simple to see that the function is broken as wrapping text refers to setting a limit to where the text on one line can't cross, and therefore adding a new line.  Calling the function a text wrapping function and intending for it to provide lines of random lengths as it does now would be somewhat of an oxymoron.  However, the documentation could be changed to naming the function minetest.reformat_text_randomly(text), and then replace the current function and leave the documentation the way it is for the minetest.wrap_text(text, charlimit).

**Comment by dsohler** [comment]
Oh, btw. here's an actually working text splitting function.

~~~~
local split_text = function (text, max_length, as_table)
  local result = {}
  local line = {}
  local as_table = as_table or false
  if text:len() <= max_length then return text end
  
  for word in text:gmatch('%S+') do
    local current_length = (table.concat(line, ' ')):len()
    local word_length = word:len() + 1
    if (current_length + word_length < max_length) then
      table.insert(line, word)
    else
      table.insert(result, table.concat(line, ' '))
      line = {}
      table.insert(line, word)
    end
  end

  table.insert(result, table.concat(line, ' '))
  if as_table == true then return result end
  return table.concat(result, '\n')
end
~~~~

Run here for see a live demonstration: https://repl.it/KZbD

**Comment by dsohler** [comment]
Okay, 2+ weeks later, still nothing new here, huh?

There are arguments for fixing/altering the functionality of the wrapping function or changing the documentation. There is even a working replacement in Lua (I bet it's easier in C++) doing what the current definition describes and even allows returning a table if wanted.

**Comment by SmallJoker** [comment]
Fixed in #6407.
